src/charm.py

```diff
@@ -1,206 +1,194 @@
+
+        if current_layer.services != new_layer.services:
+            self.container.add_layer(self._container_name, new_layer, combine=True)
+            try:
+                log.info("Pebble plan updated with new configuration, replanning")
+                self.container.replan()
+            except ChangeError as e:
+                log.error(traceback.format_exc())
+                self.unit.status = BlockedStatus("Failed to replan")
+                raise e
+                return
+
+    def _update_webhook_certs(self) -> None:
+        """Push keys and certs files into spark container"""
+        try:
+            self.container.push("/etc/webhook-certs/ca-cert.pem", self._stored.ca, make_dirs=True)
+            self.container.push(
+                "/etc/webhook-certs/server-cert.pem", self._stored.cert, make_dirs=True
+            )
+            self.container.push(
+                "/etc/webhook-certs/server-key.pem", self._stored.key, make_dirs=True
+            )
+            log.info("Pushed webhook keys and certs to spark container")
+        except (ProtocolError, PathError) as e:
+            log.error(str(e))
+            self.unit.status = BlockedStatus(str(e))
+
+    def _update_spark_container(self, event) -> None:
+        if not self.container.can_connect():
+            self.unit.status = WaitingStatus("Waiting to connect to spark container")
+            event.defer()
+            return
+
+        self.unit.status = MaintenanceStatus("Configuring Spark Charm")
+
+        self._update_webhook_certs()
+        self._update_layer()
+
+        self.unit.status = ActiveStatus()
+
+    def _on_install(self, _):
+        """Event Handler for install event."""
+        self.unit.status = MaintenanceStatus("Configuring/deploying resources")
+
+        if self.container.can_connect():
+            self._update_webhook_certs()
+
+        try:
+            self.resource_handler.apply()
+        except (ApiError, ErrorWithStatus) as e:
+            if isinstance(e, ApiError):
+                log.error(f"Applying resources failed with ApiError status code {e.status.code}")
+                self.unit.status = BlockedStatus(f"ApiError: {e.status.code}")
+            else:
+                log.info(e.msg)
+                self.unit.status = e.status
+        else:
+            self.unit.status = ActiveStatus()
+
+    def _on_spark_pebble_ready(self, event):
+        """Event Handler for spark pebble ready event."""
+        self._update_spark_container(event)
+
+    def _on_config_changed(self, event):
+        """Event Handler for config changed event."""
+        self._update_spark_container(event)
+
+    def _on_remove(self, _):
+        """Event Handler for remove event."""
+        manifests = self.resource_handler.render_manifests(force_recompute=False)
+        try:
+            delete_many(self.lightkube_client, manifests)
+            self.lightkube_client.delete(MutatingWebhookConfiguration, self._mutating_webhook_name)
```

Comment:
Having to call Client.delete() and delete_many in the same code is something that can be definitely avoided with a remove implementation in chisme. It's fine if we leave it as it is right now, I just wanted to note.